cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDocumentStoreTemplate.java

```diff
@@ -65,6 +60,57 @@ public void initialize(final AmazonDynamoDB amazonDynamoDbClient) {
         dynamoDBClient = new DynamoDB(amazonDynamoDbClient);
     }
 
+    /**
+     * Simple method for splitting a list into a list of smaller lists of the supplied length
+     * 
+     * @param list
+     * @param length
+     * @return
+     */
+    private <T> List<List<T>> split(final List<T> list, final int length) {
+        final List<List<T>> parts = new ArrayList<List<T>>();
+        final int size = list.size();
+        for (int i = 0; i < size; i += length) {
+            parts.add(new ArrayList<T>(list.subList(i, Math.min(size, i + length))));
+        }
+        return parts;
+    }
+
+    private <T extends Item> List<T> executeQuery(final KeySetQuery query, final Class<T> itemClass) {
+
+        final ItemConfiguration itemConfiguration = getItemConfiguration(itemClass);
+        final String tableName = databaseSchemaHolder.schemaName() + "." + itemConfiguration.tableName();
+
+        // max 100 keys per fetch so split in to lists of lists
+        final List<List<ItemId>> split_ids = split(new ArrayList<ItemId>(query.itemIds()), 100);
+        final List<T> fetchedItems = new ArrayList<T>();
+
+        for (final List<ItemId> ids : split_ids) {
+            final TableKeysAndAttributes keys = new TableKeysAndAttributes(tableName);
+            for (final ItemId id : ids) {
+                keys.addPrimaryKey(getPrimaryKey(id, itemConfiguration));
+            }
+            BatchGetItemOutcome outcome = dynamoDBClient.batchGetItem(keys);
+            do {
+                final List<com.amazonaws.services.dynamodbv2.document.Item> items = outcome.getTableItems().get(
+                        tableName);
+                for (final com.amazonaws.services.dynamodbv2.document.Item item : items) {
+                    fetchedItems.add(stringToItem(item.toJSON(), itemClass));
+                }
+                // Check for unprocessed keys which could happen if you exceed provisioned
+                // throughput or reach the limit on response size.
+                final Map<String, KeysAndAttributes> unprocessedKeys = outcome.getUnprocessedKeys();
+                if (outcome.getUnprocessedKeys().size() == 0) {
+                    logger.debug("No unprocessed keys found");
+                } else {
+                    logger.debug("Retrieving the unprocessed keys");
+                    outcome = dynamoDBClient.batchGetItemUnprocessed(unprocessedKeys);
```

Comment:
In the example the AWS code does not actually do anything with the outcome comparable to our line 107. So the example is correct: it will pull back all items... but doesn't do any processing on them. In our code, we will pull back all items, but our final processed (cheddar) Items will be missing some of the results.

Given there are 101 items to read in total, and the first batch read only brings back 30 due to large size of item, there will 30 items and 70 unprocessed items. In my second read of the first batch of 100, lets assume there will be 30 more items and further 40 unprocessed ones. I don't think that these 40 will be converted into cheddar Items.

Let's see what the integration test says.